from snakemake.utils import min_version

min_version("3.7.1")

cutadapt = "/home/bli/PROBer_paper_analysis/tools/cutadapt-1.10/bin/cutadapt"
PROBer = "/home/bli/software/PROBer/build/src/PROBer"

rule all:
	input: "exp/ecoli_16S_PROBer.beta"
#		expand("exp/{sample}.trimmed.bam", sample = ["DMSO_no-sel", "NPIA_no-sel", "NPIA_sel"])





rule build_ref:
	input: "ref/ecoli_rRNA.fa"
	output: touch("ref/ecoli_rRNA")
	shell:
		"{PROBer} prepare --bowtie2 {input} {output}"

rule trim_adapter:
	input: expand("data/E.coli_{{sample}}_R{mate}.fastq", mate = ["1", "2"]), cutadapt
	output: expand("data/{{sample}}_trimmed_{mate}.fastq", mate = ["1", "2"])
	log: "data/{sample}_trimmed.log"
	shell:
		"{cutadapt} -f fastq -q 17 -m 15 -a AGATCGGAAGAGCACACGTCT -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT"
		" -o {output[0]} -p {output[1]} {input[0]} {input[1]} > {log}"

rule align_reads:
	input: "ref/ecoli_rRNA", expand("data/{{sample}}_trimmed_{mate}.fastq", mate = ["1", "2"])
	output: "exp/{sample,(\w|-)+}.bam"
	threads: 10
	log: "exp/{sample}.log"
	shell:
		"bowtie2 -p {threads} -N 1 -L 15 --norc -X 700 --no-mixed --no-discordant -x {input[0]} -1 {input[1]} -2 {input[2]} 2> {log} | samtools view -b -o {output} -"

rule trim_terminals:
	input: "ref/ecoli_rRNA.fa", "exp/{sample}.bam"
	output: "exp/{sample,(\w|-)+}.trimmed.bam"
	log: "exp/{sample}.trimmed.log"
	shell:
		"scripts/trimMismatch {input} {output} > {log}"		

rule run_PROBer:
	input: "ref/ecoli_rRNA", "exp/NPIA_no-sel.trimmed.bam", "exp/DMSO_no-sel.trimmed.bam"
	output: "exp/ecoli_16S_PROBer.beta"
	threads: 10
	shell:
		"{PROBer} estimate --paired-end -p {threads} --size-selection-min 10 --size-selection-max 700 {input[0]} exp/ecoli_16S_PROBer --alignments {input[1]} {input[2]}"

rule run_PROBer_sel:
	input: "ref/ecoli_rRNA", "exp/NPIA_sel.trimmed.bam"
	output: "exp/ecoli_16S_PROBer_sel.beta"
	threads: 10
	shell:
		"{PROBer} estimate --paired-end -p {threads} --size-selection-min 10 --size-selection-max 700 {input[0]} exp/ecoli_16S_PROBer_sel --alignments {input[1]}"

